src/runtime/utils.ts

```diff
@@ -40,8 +40,15 @@ export function hasReqHeader(event: H3Event, name: string, includes: string) {
 }
 
 export function isJsonRequest(event: H3Event) {
+  // First adhere to the "Accept" header, to match what the client expects to receive.
+  if (hasReqHeader(event, "accept", "application/json")) {
+    return true;
+  }
+  if (hasReqHeader(event, "accept", "text/html")) {
+    return false;
```

Comment:
I think addition of this condition should be enough and we can keep checking application/json in fallback as before. If it is set to HTML, we early return.